File: C++/Dynamic Programming/houseRobberIII.cpp
House Robber III - LeetCode 337
  
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, int> dp;
    
    int rob(TreeNode* root) {
        if(root==NULL){
            return 0;
        }    
        
        // Memoization
        if(dp[root])
            return dp[root];
        
        // Root Data included
        int inc = root->val;
        
        if(root->left)
            inc += rob(root->left->left) + rob(root->left->right);
        
        if(root->right)
            inc += rob(root->right->left) + rob(root->right->right);
        
        int exc = rob(root->left) + rob(root->right);
        
        dp[root] = max(inc, exc);
        
        return dp[root];
    }
};

File: C++/Dynamic Programming/Revision 2025/HouseRobberIII.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        
        # [Always return the value - [WithNode, WithoutNode]]
        def dfs(root):
            if root is None:
                return [0, 0]
            
            leftPair = dfs(root.left)       #[withLeftNode, withoutLeftNode]
            rightPair = dfs(root.right)     #[wightRightNode, withoutRightNode]

            # If root is include we can only take without values from left and right
            withRoot = root.val + leftPair[1] + rightPair[1]

            # If root is not included. There is no restriction now.
            # Take the max from left and right
            withoutRoot = max(leftPair) + max(rightPair)

            return [withRoot, withoutRoot]
        
        withRoot, withoutRoot = dfs(root)
        return max(withRoot, withoutRoot)
        
